src/modules/track_model/src/track_model.cc

```diff
@@ -1 +1,339 @@
+
+                train_head_[i] = temp_distance - d_traveled;
+
+                //loop to account for length of train
+                types::Meters current_length = blocks_[current_block].length - train_head_[i];
+
+                while (current_length < length_of_train)
+                {
+                    //did we hit the yard?
+                    if (track_path_[j].block == 0)
+                    {
+                        break;
+                    }
+
+                    //decrement to next block
+                    j--;
+                    current_length += track_path_[j].length;
+
+                    //update occupany
+                    types::BlockId next_block = track_path_[j].block;
+                    occupied_train_blocks_[i].push_back(next_block);
+                    blocks_[next_block].occupied = 1;
+
+                }
+
+                break;
+            }
+
+        }
+
+        // Check if the current block has a station and update deboarding
+        types::BlockId possible_station_block = current_train_block_[i];
+        if (blocks_[possible_station_block].has_station == 1)
+        {
+            uint16_t traindeb = trains_[i]->GetPassengersDeboarding();
+            SetPassengersDeboarding(i, traindeb);
+        }
+    }
+
+}
+
+
+
+types::Error SoftwareTrackModel::SetSwitchState(const types::BlockId block, const bool switched)
+{
+    if (blocks_.size() > block && block > 0 && blocks_[block].has_switch == 1)
+    {
+        blocks_[block].switched = switched;
+
+        return types::ERROR_NONE;
+    }
+    else
+    {
+        return types::ERROR_INVALID_BLOCK;
+    }
+
+}
+
+types::Error SoftwareTrackModel::SetCrossingState(const types::BlockId block, const bool closed)
+{
+    return types::Error{ };
+}
+
+types::Error SoftwareTrackModel::SetRedTrafficLight(const types::BlockId block, const bool on)
+{
+    return types::ERROR_NONE;
+}
+
+types::Error SoftwareTrackModel::SetYellowTrafficLight(const types::BlockId block, const bool on)
+{
+    return types::ERROR_NONE;
+}
+
+types::Error SoftwareTrackModel::SetGreenTrafficLight(const types::BlockId block, const bool on)
+{
+    return types::ERROR_NONE;
+}
+
+types::Error SoftwareTrackModel::SetCommandedSpeed(const types::BlockId block, const types::MetersPerSecond speed)
+{
+    if (blocks_.size() > block && block > 0)
+    {
+        for (int i = 0; i < trains_.size(); i++)
+        {
+            for (int j = 0; j < occupied_train_blocks_[i].size(); j++)
+            {
+                if (occupied_train_blocks_[i][j] == block)
+                {
+                    trains_[i]->SetCommandedSpeed(speed);
+                }
+            }
+        }
+
+        return types::ERROR_NONE;
+    }
+    else
+    {
+        return types::ERROR_INVALID_BLOCK;
+    }
+}
+
+types::Error SoftwareTrackModel::SetAuthority(const types::BlockId block, const types::Blocks authority)
+{
+    if (blocks_.size() > block && block > 0)
```

Comment:
Code too heavily nested

Reply:
there is no way to make this less nested without causing a segfault as far as i can tell